### ISSUE TITLE
grey_webSetText fails when there's no title or value

### ISSUE THREAD

**Issue by bootstraponline** [google/EarlGrey#211]
Given the following input:

> <input class="ic-Input" type="text" name="pseudonym_session[unique_id]" value="" placeholder="Email">

Matching on title or value will fail.

```
$x('//input[@title="Email" or @value="Email"]')
> []
```

Place holder matching works but it's not unique.

```
$x('//input[@title="Email" or @value="Email" or @placeholder="Email"]')
> [<input class=​"ic-Input" type=​"text" name=​"pseudonym_session[unique_id]​" value placeholder=​"Email">​, <input class=​"ic-Input" type=​"text" name=​"pseudonym_session[unique_id_forgot]​" value placeholder=​"Email" id=​"pseudonym_session_unique_id_forgot">​]
```

I think we need to allow matching on arbitrary XPath selectors since there's no way to control the content of the webview.

**Comment by bootstraponline** [comment]
In Espresso, I can login via arbitrary CSS selectors or XPath.

```java
// Espresso
String EMAIL = "input[type=\"email\"]";
String PASSWORD = "input[type=\"password\"]";

onWebView()
  .withElement(findElement(Locator.CSS_SELECTOR, EMAIL))
  .perform(webKeys("some email"))
  .withElement(findElement(Locator.CSS_SELECTOR, PASSWORD))
  .perform(webKeys("password"))
```

**Comment by bootstraponline** [comment]
https://gist.github.com/tirodkar/cf4a72d98264545749b063e79838fa03 looks like a good work around until the new webview APIs are ready.

**Comment by khandpur** [comment]
Would adding a css selector as a matcher address this issue? If so, we could rephrase this to sound like a feature request and add the example above as a test for the feature.

**Comment by bootstraponline** [comment]
> Would adding a css selector as a matcher address this issue?

Yes.

**Comment by bootstraponline** [comment]
some notes since I've started looking into this:

Espresso serializes HTML elements to a JSON [ElementReference](
https://github.com/bootstraponline/platform_frameworks_testing/blob/android-support-test/espresso/web/src/main/java/android/support/test/espresso/web/model/ElementReference.java)

That looks like:
> {"ELEMENT":":wdc:1490646488594"}

which is [standard WebDriver](https://github.com/SeleniumHQ/selenium/blob/0fd840096fec20363daef3161a825e937b8bdc7e/javascript/atoms/inject.js#L442).

I haven't yet figured out how to create references from a located HTML element.

**Comment by bootstraponline** [comment]
Here's the code for my port of Espresso's WebView logic to Swift & EarlGrey.

- https://github.com/instructure/instructure-ios/commit/d0dcc75aee6799ea7bddcd6ae492aa5ffe79ac57

<img width="595" alt="css_selector_victory" src="https://cloud.githubusercontent.com/assets/1173057/24472615/518b4bf0-1494-11e7-8379-9c87f9e084b2.png">

Finding by CSS selector is working. It's very much a proof of concept with most of the APIs not yet implemented. The approach is verified as working though.

**Comment by bootstraponline** [comment]
I pushed an update which adds element reference support & webKeys (typing text).

- https://github.com/instructure/instructure-ios/commit/25f82f0bb1e2f9f493a76e5406c1fb6d4d9f0d58

**Comment by drzajwo** [comment]
> I pushed an update which adds element reference support & webKeys (typing text).
> 
> * https://github.com/instructure/instructure-ios/commit/25f82f0bb1e2f9f493a76e5406c1fb6d4d9f0d58
> 
> Example usage:
> 
> ```swift
> private let EMAIL_FIELD_CSS = "input[name=\"email\"]";
> if let emailElement = DriverAtoms.findElement(locator: Locator.CSS_SELECTOR, value: EMAIL_FIELD_CSS) {
>   DriverAtoms.webKeys(element: emailElement, value: "email@example.com")
> }
> ```

As we spoke on Slack, could you please update the links? Maybe I'll try to implement your solution which seems to be very interesting ;)

**Comment by bootstraponline** [comment]
https://github.com/bootstraponline/SoGrey/blob/master/SoGrey/SoGrey/DriverAtoms.swift

**Comment by tirodkar** [comment]
Do you have a matcher we could add in EG 2 as well? @bootstraponline

**Comment by bootstraponline** [comment]
I completed the experimental prototype for EarlGrey 1. I'm not sure if it works for EG2. I expect it'd need some stabilization before being mainlined. XCUITest solved the simple webview use case I cared about.

I think a larger project to port espresso web for real to EarlGrey would be nice.
